### PR TITLE
FIX: Implementation of analytical integral of linearly interpolated Functions

### PR DESCRIPTION
**Pull request by giovaniceotto**
## Pull request type

Please check the type of change your PR introduces:
- [x] Code base additions (bugfix, features)

## Pull request checklist

Please check if your PR fulfills the following requirements, depending on the type of PR:

- Code base additions (for bug fixes / features):

  - [ ] Tests for the changes have been added
  - [x] Docs have been reviewed and added / updated if needed
  - [x] Lint (`black rocketpy`) has passed locally and any fixes were made
  - [x] All tests (`pytest --runslow`) have passed locally

## What is the current behavior?

See #331.

## What is the new behavior?

Issue #331 is fixed by implementing a correct analytical integral of linearly interpolated Functions.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

Refactoring the Flight class is long overdue... I am not comfortable with the size (number of lines of code) of the integral method, but creating "sub" functions goes against the current organization of the Function class.

### CODE REVIEW

**Comment by giovaniceotto** [comment]
@phmbressan, please ignore the commit by Lint-Action

File: rocketpy/Function.py [hunk 1]
```diff
@@ -2227,11 +2231,31 @@ def integral(self, a, b, numerical=False):
...
         elif self.__interpolation__ == "linear" and numerical is False:
-            return np.trapz(self.source[:, 1], x=self.source[:, 0])
+            # Integrate from a to b using np.trapz
+            xData = self.source[:, 0]
+            yData = self.source[:, 1]
+            # Get data in interval
+            xIntegrationData = xData[(xData >= a) & (xData <= b)]
+            yIntegrationData = yData[(xData >= a) & (xData <= b)]
+            # Add integration limits to data
+            if self.__extrapolation__ == "zero":
```
Comment (phmbressan) [inline rocketpy/Function.py:2241]: The idea of adding the integration limits to the data was interesting for consistency, however it introduces new spurious results when integrating with `np.trapz`. See below:

```
>>> from rocketpy import Function
>>> f = Function([(0,0), (1,2), (10,20)], interpolation="linear", extrapolation="zero")
>>> f.integral(0,11) # "exact" integral
110.0  # should be 100.0
>>> f.integral(0,11,True)
100.00083463213139
```

The problem is that the integral limit in the zone of zero extrapolation is causing the trapezoid rule to interpret one more region as a trapezoid whose lesser base is 0 (i.e. a triangle).

I do not have a nice fix for the problem, however, if the integration limit is not added to the dataset, the function gives the correct value.

File: rocketpy/Function.py [hunk 2]
```diff
@@ -2227,11 +2231,31 @@ def integral(self, a, b, numerical=False):
...
+                if a >= xData[0]:
+                    xIntegrationData = np.concatenate(([a], xIntegrationData))
+                    yIntegrationData = np.concatenate(([self(a)], yIntegrationData))
+                if b <= yData[-1]:
```
Comment (giovaniceotto) [inline rocketpy/Function.py:2245]:
```suggestion
                if b <= xData[-1]:
```

File: rocketpy/Function.py (hunk 1 above)
Comment (giovaniceotto) [reply rocketpy/Function.py:2241]: A typo in line 2245, where `yData` should be `xData` caused the problem. I have fixed it in the latest commit. Therefore, I am requesting your review once again @phmbressan

**Review by phmbressan — APPROVED** [review]
Excellent. The method now seems to be correct.

**Pull request merged**